File: src/main/java/org/noxet/noxetserver/playerstate/properties/PSPAdvancementCriteria.java
package org.noxet.noxetserver.playerstate.properties;

import org.bukkit.advancement.Advancement;
import org.bukkit.advancement.AdvancementProgress;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.Player;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.playerstate.PlayerStateProperty;

import java.util.Iterator;

import static org.noxet.noxetserver.playerstate.PlayerState.getAdvancementCriteriaList;

public class PSPAdvancementCriteria implements PlayerStateProperty<String[]> {
    @Override
    public String getConfigName() {
        return "advancement_criteria";
    }

    @Override
    public String[] getDefaultSerializedProperty() {
        return new String[0];
    }

    @Override
    public String[] getSerializedPropertyFromPlayer(Player player) {
        return getAdvancementCriteriaList(player).toArray(new String[0]);
    }

    @Override
    public void restoreProperty(Player player, String[] restoreCriteriaList) {
        Iterator<Advancement> advancementIterator = NoxetServer.getPlugin().getServer().advancementIterator();

        while(advancementIterator.hasNext()) {
            AdvancementProgress progress = player.getAdvancementProgress(advancementIterator.next());

            for(String criteria : progress.getRemainingCriteria()) // Check all not yet rewarded criteria
                for(String hasCriteria : restoreCriteriaList) // Loop all criteria which should be set
                    if(criteria.equals(hasCriteria)) // Is a criteria which should be set, not set?
                        progress.awardCriteria(criteria);

            for(String existingCriteria : progress.getAwardedCriteria()) {
                boolean revokeCritera = true;

                for(String hasCriteria : restoreCriteriaList)
                    if(existingCriteria.equals(hasCriteria)) {
                        revokeCritera = false;
                        break;
                    }

                if(revokeCritera)
                    progress.revokeCriteria(existingCriteria);
            }
        }
    }

    @Override
    public String[] getValueFromConfig(ConfigurationSection config) {
        return config.getStringList(getConfigName()).toArray(new String[0]);
    }
}

File: src/main/java/org/noxet/noxetserver/playerstate/properties/PSPAdvancementCriteria.scala
package org.noxet.noxetserver.playerstate.properties

import org.bukkit.advancement.{Advancement, AdvancementProgress}
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.entity.Player
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.playerstate.PlayerState.getAdvancementCriteriaList
import org.noxet.noxetserver.playerstate.PlayerStateProperty

import scala.jdk.CollectionConverters.*

object PSPAdvancementCriteria extends PlayerStateProperty[Array[String]]:
  override def getConfigName: String = "advancement_criteria"

  override def getDefaultSerializedProperty: Array[String] = Array.empty

  override def getSerializedPropertyFromPlayer(player: Player): Array[String] =
    getAdvancementCriteriaList(player).toArray

  override def restoreProperty(player: Player, restoreCriteriaList: Array[String]): Unit =
    NoxetServer.getPlugin.getServer.advancementIterator
      .map(player.getAdvancementProgress)
      .foreach(p =>
        p.getRemainingCriteria.asScala.toList
          .intersect(restoreCriteriaList)
          .foreach(p.awardCriteria)

        p.getAwardedCriteria.asScala.toList
          .filter(_.intersect(restoreCriteriaList).isEmpty)
          .foreach(p.revokeCriteria)
      )

  override def getValueFromConfig(config: ConfigurationSection): Array[String] =
    config.getStringList(getConfigName).asScala.toArray